### PR TITLE
[ENH] Extend extract_b0 capabilities

### PR DESCRIPTION
**Pull request by boiteaclou**
This PR contains modifications to :

- Extract b0 : added more extraction options, such as extracting b0s from clusters of b0s in the data (mean or first b0 of each cluster). Also reformatted the code to enable lazyloading. (Does not modify previous behavior of the script, fully back-compatible cli)
- Prepare Topup : TBD
- Prepare Eddy : TBD

### CODE REVIEW

**Comment by boiteaclou** [comment]
First question about the prepare_topup cli. Right now, we ask for the full forward phase encoded dwi file/bval/bvec. Two options for the new script :
- Keep this behavior and add an argument to allow passing only b0s files. This however make the cli non back-compatible, but makes the most sense since the b0s to use should have been previously extracted using `scil_extract_b0`
- Replace passage of full forward phase encoded dwi data and only allow by passage of b0s. This also makes the cli non back-compatible
- Keep it as is and add the options of `scil_extract_b0` to the cli to handle b0 extraction from the full forward phase encoded dwi

**Comment by arnaudbore** [comment]
Quick note: scil_prepare_topup extracts the b0s from the dwi and compute the mean and does the same for the rev_b0. My first guess would be to allow either dwi input or b0 for each Phase Encoding direction and the rest of the code should be as is. I find your modifications pretty heavy compare to what I had in mind but maybe this is mandatory. We should have a discussion with @GuillaumeTh so everybody is on the same page here before moving forward.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Review by arnaudbore — APPROVED** [review]

File: scilpy/image/utils.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def count_non_zero_voxels(image):
     return nb_voxels
 
 
-def volume_iterator(img, blocksize=1):
+def volume_iterator(img, blocksize=1, start=0, end=0):
```
Comment (GuillaumeTh) [inline scilpy/image/utils.py:37]: Please add documentation for start and end.

File: scripts/scil_extract_b0.py [hunk 2]
```diff
@@ -39,46 +40,45 @@ def _build_arg_parser():
...
-                        'diffusion weighting.')
+                        'diffusion weighting. [%(default)s]')
 
     group = p.add_mutually_exclusive_group()
     group.add_argument('--all', action='store_true',
                        help='Extract all b0. Index number will be appended to '
                             'the output file.')
     group.add_argument('--mean', action='store_true', help='Extract mean b0.')
+    group.add_argument('--cluster-mean', action='store_true',
+                       help='Extracts mean of each continuous cluster of b0s')
```
Comment (GuillaumeTh) [inline scripts/scil_extract_b0.py:51]: Extracts -> Extract and . at the end.

File: scripts/scil_extract_b0.py [hunk 3]
```diff
@@ -39,46 +40,45 @@ def _build_arg_parser():
...
+    group.add_argument('--cluster-first', action='store_true',
+                       help='Extracts first b0 of each '
```
Comment (GuillaumeTh) [inline scripts/scil_extract_b0.py:53]: Extracts -> Extract and . at the end.

File: scripts/scil_extract_b0.py [hunk 4]
```diff
@@ -39,46 +40,45 @@ def _build_arg_parser():
...
+                            'continuous cluster of b0s')
+
+    p.add_argument('--block-size', '-s',
+                   metavar='INT', type=int,
+                   help='Loads the data using this block size. '
```
Comment (GuillaumeTh) [inline scripts/scil_extract_b0.py:58]: Loads -> Load

**Review by GuillaumeTh — APPROVED** [review]
I just checked the code. A small comment otherwise looks good.

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Pull request merged**